tools/functional-tester/etcd-agent/agent.go

```diff
@@ -80,8 +81,23 @@ func (a *Agent) stop() error {
 	if a.state != stateStarted {
 		return nil
 	}
-	err := a.cmd.Process.Kill()
-	if err != nil {
+
+	if err := syscall.Kill(a.cmd.Process.Pid, syscall.SIGTERM); err != nil {
+		return err
+	}
+	time.Sleep(time.Second)
+
+	if err := syscall.Kill(a.cmd.Process.Pid, syscall.SIGINT); err != nil {
+		return err
+	}
+	time.Sleep(time.Second)
+
+	if err := syscall.Kill(a.cmd.Process.Pid, syscall.SIGHUP); err != nil {
+		return err
+	}
+	time.Sleep(time.Second)
+
+	if err := a.cmd.Process.Kill(); err != nil {
```

Comment:
I think you can get rid of all the `Sleep`s. The final SIGKILL, if any, can happen after timing out on the Wait:
```
go func() {
_, err := a.cmd.Process.Wait()
errc <- err
close(errc)
}()
select {
case <-time.After(5 * time.Second):
a.cmd.Process.Kill()
case err := <-errc:
...
}
err <-errc
...
```